### DISCUSSION TITLE
Multiple C# projects in the same root directory?

### DISCUSSION THREAD

**Discussion by geekox86** [dotnet/msbuild#6008] [Q&A]
I have a root directory with multiple C# projects (*.csproj).
I need MSBuild to filter the C# source files by root namespace for each project, since they share the same root directory.
Otherwise, I will get `error CS0017: Program has more than one entry point defined` when I try to run a project using `dotnet run --project Project1.csproj`.
My project folder structure is organized based on features, and I want all source files related to a feature from all projects (e.g. server, client, test) to be inside the feature specific sub-folder.

Sample solution structure:
```
|_ Project1.csproj
|_ Project2.csproj
|_ Program1.cs (this contains main for Project1 under root namespace P1)
|_ Program2.cs (this contains main for Project2 under root namespace P2)
|_ Feature1 (this feature belongs to Project1 under root namespace P1)
|  |_ SomeClass1.cs
|  |_ SomeClass1.test.cs
|_ Feature2 (this feature belongs to Project2 under root namespace P2)
   |_ SomeClass2.cs
   |_ SomeClass2.test.cs
```

Now, how can I run/build Project1.csproj such that MSBuild will only include Program1.cs and SomeClass1.cs. Similar thing with Project2.csproj with Program2.cs and SomeClass2.cs. Can I ask MSBuild to include only files/types based on their root namespace?

Any idea how to go about this?

**Comment by drewnoakes** [comment]
To fix the multiple entry point issue, in `Project1.csproj` you could add:

```xml
<ItemGroup>
    <Compile Remove="Program*.cs" />
    <Compile Include="Program1.cs" />
</ItemGroup>
```

It does seem like this would be easier to achieve with multiple projects though.

**Reply by KalleOlaviNiemitalo** [reply]
The `Exclude` attribute only prevents items from being added by an `Include` attribute in the same element: <https://github.com/dotnet/msbuild/blob/39993bd9d02917993b6147a9973dc3aa6e9b00c7/src/Build/Evaluation/ProjectParser.cs#L292-L293>

You might instead need the `Remove` attribute, or properties listed in [Default item inclusion properties](https://docs.microsoft.com/en-us/dotnet/core/project-sdk/msbuild-props#default-item-inclusion-properties).

In addition, you should set the `BaseIntermediateOutputPath` property so that the files generated from these projects or added by NuGet do not conflict.

**Reply by drewnoakes** [reply]
I see I also used `Exclude` instead of `Include`. I've updated the snippet.

**Comment by Nirmal4G** [comment]
If you're using `.NET SDK` then set `EnableDefaultItems` property to `false` and manually include them by a pattern from which you can isolate them in the includes.

Example:
```xml
<PropertyGroup>
    <EnableDefaultItems>False</EnableDefaultItems>
</PropertyGroup>

<!-- Project 1 -->
<ItemGroup>
    <Compile Include="*/**/*_P1.cs" Exclude="$(DefaultExcludesInProjectFolder);$(DefaultItemExcludes)" />
</ItemGroup>

<!-- Project 2 -->
<ItemGroup>
    <Compile Include="*/**/*_P2.cs" Exclude="$(DefaultExcludesInProjectFolder);$(DefaultItemExcludes)" />
</ItemGroup>
```

For more info, see [default item inclusions](https://docs.microsoft.com/dotnet/core/project-sdk/msbuild-props#default-item-inclusion-properties) in `Microsoft.NET.Sdk`.
